File: core/src/main/java/emprestes/ds/domain/data/DynamicArray.java
package emprestes.ds.domain.data;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

/**
 * Simple resizable array structure.
 *
 * @param <T> stored element type
 */
public class DynamicArray<T> {

    private static final int DEFAULT_CAPACITY = 10;

    private Object[] values = new Object[DEFAULT_CAPACITY];
    private int size;

    /**
     * Appends a value at the end of the array.
     *
     * @param value value to append
     * @return current array instance
     */
    public DynamicArray<T> add(T value) {
        if (value == null) {
            return this;
        }
        ensureCapacity(size + 1);
        values[size++] = value;
        return this;
    }

    /**
     * Returns the value at the given index.
     *
     * @param index zero-based index
     * @return stored value
     */
    public T get(int index) {
        validateIndex(index);
        return elementAt(index);
    }

    /**
     * Replaces a value at the given index.
     *
     * @param index zero-based index
     * @param value new value
     * @return current array instance
     */
    public DynamicArray<T> set(int index, T value) {
        validateIndex(index);
        values[index] = value;
        return this;
    }

    /**
     * Removes and returns the value at the given index.
     *
     * @param index zero-based index
     * @return removed value
     */
    public T removeAt(int index) {
        validateIndex(index);
        var removed = elementAt(index);
        int moveCount = size - index - 1;
        if (moveCount > 0) {
            System.arraycopy(values, index + 1, values, index, moveCount);
        }
        values[--size] = null;
        return removed;
    }

    /**
     * Checks whether the array contains the value.
     *
     * @param value value to search
     * @return {@code true} when present
     */
    public boolean contains(T value) {
        for (int i = 0; i < size; i++) {
            if (Objects.equals(values[i], value)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns the current number of elements.
     *
     * @return array size
     */
    public int size() {
        return size;
    }

    /**
     * Indicates whether no values are stored.
     *
     * @return {@code true} when empty
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * Returns the current array values as an immutable list.
     *
     * @return list snapshot
     */
    public List<T> toList() {
        var copy = Arrays.copyOf(values, size);
        @SuppressWarnings("unchecked")
        var typed = (T[]) copy;
        return List.of(typed);
    }

    /**
     * Expands internal capacity to fit at least minCapacity elements.
     *
     * @param minCapacity required minimum capacity
     */
    private void ensureCapacity(int minCapacity) {
        if (minCapacity <= values.length) {
            return;
        }
        int nextCapacity = values.length * 2;
        while (nextCapacity < minCapacity) {
            nextCapacity = nextCapacity * 2;
        }
        values = Arrays.copyOf(values, nextCapacity);
    }

    /**
     * Returns the typed element at a valid index.
     *
     * @param index validated index
     * @return element value
     */
    @SuppressWarnings("unchecked")
    private T elementAt(int index) {
        return (T) values[index];
    }

    /**
     * Validates index bounds for the current size.
     *
     * @param index index to validate
     */
    private void validateIndex(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
        }
    }
}

File: core/src/main/kotlin/emprestes/ds/array/kotlin/DynamicArray.kt
package emprestes.ds.array.kotlin

/**
 * Resizable array implemented on top of a contiguous block of references.
 *
 * The structure separates two concepts that are often confused: [size] is the
 * number of visible elements, while [capacity] is the number of slots reserved
 * in the internal storage. When an insertion does not fit, the capacity doubles
 * and the elements are copied into a new block. Therefore, appending costs
 * amortized O(1), although the operation that grows the array costs O(n).
 *
 * This implementation accepts null values and does not use [MutableList]
 * internally, so the mechanics of the data structure remain explicit.
 *
 * @param E type of the stored elements.
 * @property initialCapacity initial number of reserved slots; must be positive.
 */
class DynamicArray<E>(initialCapacity: Int = DEFAULT_CAPACITY) : Iterable<E> {
    private var elements: Array<Any?>

    /** Number of elements currently stored. */
    var size: Int = 0
        private set

    /** Number of elements that fit without another reallocation. */
    val capacity: Int
        get() = elements.size

    /** Whether the structure contains no elements. */
    val isEmpty: Boolean
        get() = size == 0

    init {
        require(initialCapacity > 0) { "initialCapacity must be greater than zero" }
        elements = arrayOfNulls(initialCapacity)
    }

    /**
     * Appends [element] to the end.
     *
     * Complexity: amortized O(1), and O(n) in the worst case when storage grows.
     */
    fun add(element: E) {
        ensureCapacity(size + 1)
        elements[size++] = element
    }

    /**
     * Inserts [element] at [index], shifting successors one position to the right.
     *
     * [index] may equal [size], in which case this operation is equivalent to [add].
     * Complexity: O(n).
     */
    fun add(index: Int, element: E) {
        checkPositionIndex(index)
        ensureCapacity(size + 1)
        elements.copyInto(elements, destinationOffset = index + 1, startIndex = index, endIndex = size)
        elements[index] = element
        size++
    }

    /** Returns the element at [index]. Complexity: O(1). */
    operator fun get(index: Int): E {
        checkElementIndex(index)
        return elementAt(index)
    }

    /** Replaces and returns the previous value at [index]. Complexity: O(1). */
    operator fun set(index: Int, element: E): E {
        checkElementIndex(index)
        val previous = elementAt(index)
        elements[index] = element
        return previous
    }

    /**
     * Removes and returns the element at [index], closing the gap left behind.
     *
     * Complexity: O(n), because successors must be shifted.
     */
    fun removeAt(index: Int): E {
        checkElementIndex(index)
        val removed = elementAt(index)
        val lastIndex = size - 1
        if (index < lastIndex) {
            elements.copyInto(elements, destinationOffset = index, startIndex = index + 1, endIndex = size)
        }
        elements[lastIndex] = null
        size--
        return removed
    }

    /** Removes every element without shrinking reserved capacity. Complexity: O(n). */
    fun clear() {
        elements.fill(null, fromIndex = 0, toIndex = size)
        size = 0
    }

    /** Creates a read-only list with the current elements. Complexity: O(n). */
    fun toList(): List<E> = List(size) { elementAt(it) }

    /** Iterates in index order, from zero through [size] - 1. */
    override fun iterator(): Iterator<E> = object : Iterator<E> {
        private var cursor = 0

        override fun hasNext(): Boolean = cursor < size

        override fun next(): E {
            if (!hasNext()) throw NoSuchElementException()
            return elementAt(cursor++)
        }
    }

    private fun ensureCapacity(requiredCapacity: Int) {
        if (requiredCapacity <= capacity) return
        var newCapacity = capacity
        while (newCapacity < requiredCapacity) newCapacity *= GROWTH_FACTOR
        elements = elements.copyOf(newCapacity)
    }

    @Suppress("UNCHECKED_CAST")
    private fun elementAt(index: Int): E = elements[index] as E

    private fun checkElementIndex(index: Int) {
        if (index !in 0 until size) {
            throw IndexOutOfBoundsException("index=$index, size=$size")
        }
    }

    private fun checkPositionIndex(index: Int) {
        if (index !in 0..size) {
            throw IndexOutOfBoundsException("index=$index, size=$size")
        }
    }

    private companion object {
        const val DEFAULT_CAPACITY = 10
        const val GROWTH_FACTOR = 2
    }
}
